File: daily_challenges/3494.py
# Tc = O(m * n)
# Sc = O(n)

from typing import List

class Solution:
    def minTime(self, skill: List[int], mana: List[int]) -> int:
        n, m = len(skill), len(mana)
        finishTime = [0] * n  # finishTime[i] = when wizard i finishes current potion

        for j in range(m):
            finishTime[0] += mana[j] * skill[0]

            for i in range(1, n):
                finishTime[i] = max(finishTime[i], finishTime[i - 1]) + mana[j] * skill[i]

            for i in range(n - 1, 0, -1):
                finishTime[i - 1] = finishTime[i] - mana[j] * skill[i]

        return finishTime[n - 1]


            

File: daily_challenge_leetcode/3494.js
/**
 * @param {number[]} skill
 * @param {number[]} mana
 * @return {number}
 */
var minTime = function (skill, mana) {
    const n = skill.length
    const m = mana.length
    const finishTime = new Array(n).fill(0)
    for (let j = 0; j < m; j++) {
        finishTime[0] += mana[j] * skill[0]
        for (let i = 1; i < n; i++) {
            finishTime[i] = Math.max(finishTime[i], finishTime[i - 1]) + mana[j] * skill[i]
        }
        for (let i = n - 1; i > 0; i--) {
            finishTime[i - 1] = finishTime[i] - mana[j] * skill[i]
        }
    }
    return finishTime[n - 1]
};